File: torarvid/src/day01.rs
use crate::util::file_by_lines_as_i32;

pub fn run() {
    let numbers = file_by_lines_as_i32("day01.txt");
    let mut count = 0;
    for i in 1..numbers.len() {
        if numbers[i] > numbers[i - 1] {
            count += 1;
        }
    }
    println!("Part 1: {}", count);

    // Part 2
    let mut group_sum: Option<i32> = None;
    count = 0;
    for i in 2..numbers.len() {
        let new_group_sum = numbers[i - 2] + numbers[i - 1] + numbers[i];
        if group_sum.is_some() && group_sum.unwrap() < new_group_sum {
            count += 1;
        }
        group_sum = Some(new_group_sum);
    }
    println!("Part 2: {}", count);
}

File: torarvid/day01.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const util = @import("util.zig");

pub fn part1(input: []const []const u8) !void {
    var password: u16 = 0;
    var position: i9 = 50;
    for (input) |line| {
        var adjustment: i12 = 1;
        if (line[0] == 'L') {
            adjustment = -1;
        }
        const element = try std.fmt.parseInt(u11, line[1..line.len], 10);
        adjustment *= element;
        const new_position = @mod(position + adjustment, 100);
        position = @intCast(new_position);
        if (position == 0) {
            password += 1;
        }
    }

    std.debug.print("Part 1: {d}\n", .{password});
}

pub fn part2(input: []const []const u8) !void {
    var password: u16 = 0;
    var position: i9 = 50;
    for (input) |line| {
        var adjustment: i12 = 1;
        if (line[0] == 'L') {
            adjustment = -1;
        }
        const element = try std.fmt.parseInt(u11, line[1..line.len], 10);
        for (0..element) |_| {
            const new_position = @mod(position + adjustment, 100);
            position = @intCast(new_position);
            if (position == 0) {
                password += 1;
            }
        }
    }

    std.debug.print("Part 2: {d}\n", .{password});
}

pub fn run(input: []const []const u8, allocator: Allocator) !void {
    try part1(input);
    try part2(input);

    for (input) |line| {
        allocator.free(line);
    }
    allocator.free(input);
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const input = try util.file_as_strings("inputs/day01.txt", allocator);

    try run(input, allocator);
}

test "day01" {
    const allocator = std.testing.allocator;
    const raw =
        \\L68
        \\L30
        \\R48
        \\L5
        \\R60
        \\L55
        \\L1
        \\L99
        \\R14
        \\L82
    ;
    var reader = std.Io.Reader.fixed(raw);

    const input = try util.read_as_strings(&reader, allocator);
    try run(input, allocator);
}
